File: ios/SpentyAI/SpentyAI/Features/PastInsights/PastInsightsRepository.swift
import Foundation

// MARK: - Request Models

struct CreateTaxSummaryRequest: Codable {
    let name: String
    let dateFrom: String
    let dateTo: String
    let emailAddress: String
}

struct UpdateTransactionRequest: Codable {
    let date: String?
    let description: String?
    let amount: Double?
    let transactionType: String?
    let category: String?
}

struct AddTransactionRequest: Codable {
    let date: String
    let description: String
    let amount: Double
    let transactionType: String
    let category: String?
}

// MARK: - Response Models

struct TaxSummaryListResponse: Codable {
    let summaries: [TaxSummary]
}

struct TaxSummaryDetailResponse: Codable {
    let summary: TaxSummary
    let transactions: [TaxSummaryTransaction]
}

struct EmailOption: Codable {
    let email: String
    let provider: String
}

struct AvailableEmailsResponse: Codable {
    let emails: [EmailOption]
}

struct GenerateResponse: Codable {
    let message: String?
    let id: String?
    let dateFrom: String?
    let dateTo: String?
    let totalIncome: Double?
    let totalExpenses: Double?
    let net: Double?
    let transactionCount: Int?
}

// MARK: - Repository

final class PastInsightsRepository {

    static let shared = PastInsightsRepository()
    private init() {}

    // MARK: - Summaries

    func getSummaries() async throws -> [TaxSummary] {
        let response: TaxSummaryListResponse = try await APIClient.shared.get(APIEndpoints.taxSummary)
        return response.summaries
    }

    func getSummaryDetail(id: String) async throws -> TaxSummaryDetailResponse {
        try await APIClient.shared.get(APIEndpoints.taxSummaryDetail(id))
    }

    func createSummary(_ request: CreateTaxSummaryRequest) async throws -> TaxSummary {
        try await APIClient.shared.post(APIEndpoints.taxSummary, body: request)
    }

    func generateSummary(dateFrom: String, dateTo: String) async throws -> GenerateResponse {
        let path = APIEndpoints.taxSummaryGenerate + "?date_from=\(dateFrom)&date_to=\(dateTo)"
        return try await APIClient.shared.get(path)
    }

    func deleteSummary(id: String) async throws {
        let _: EmptyResponse = try await APIClient.shared.delete(APIEndpoints.taxSummaryDetail(id))
    }

    func getAvailableEmails() async throws -> [EmailOption] {
        let response: AvailableEmailsResponse = try await APIClient.shared.get(APIEndpoints.taxSummaryAvailableEmails)
        return response.emails
    }

    // MARK: - Transactions

    func addTransaction(summaryId: String, _ request: AddTransactionRequest) async throws -> TaxSummaryTransaction {
        try await APIClient.shared.post(APIEndpoints.taxSummaryTransactions(summaryId), body: request)
    }

    func updateTransaction(summaryId: String, txnId: String, _ request: UpdateTransactionRequest) async throws -> TaxSummaryTransaction {
        try await APIClient.shared.put(APIEndpoints.taxSummaryTransaction(summaryId, txnId), body: request)
    }

    func deleteTransaction(summaryId: String, txnId: String) async throws {
        let _: EmptyResponse = try await APIClient.shared.delete(APIEndpoints.taxSummaryTransaction(summaryId, txnId))
    }

    // MARK: - Export / Download

    func exportCSV(summaryId: String) async throws -> Data {
        try await APIClient.shared.getRaw(APIEndpoints.taxSummaryExport(summaryId))
    }

    func downloadCSV(summaryId: String) async throws -> Data {
        // Backend download endpoint also returns CSV, not PDF
        try await APIClient.shared.getRaw(APIEndpoints.taxSummaryDownload(summaryId))
    }
}

File: android-native/app/src/main/java/com/spentyai/app/features/pastinsights/PastInsightsRepository.kt
package com.spentyai.app.features.pastinsights

import com.spentyai.app.core.network.ApiClient
import com.spentyai.app.core.network.ApiResult
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

// --- Models ---

@Serializable
data class PastInsightSummary(
    val id: String = "",
    val name: String? = null,
    @SerialName("date_from") val dateFrom: String? = null,
    @SerialName("date_to") val dateTo: String? = null,
    val status: String? = null,
    @SerialName("total_income") val totalIncome: Double? = null,
    @SerialName("total_expense") val totalExpense: Double? = null,
    val net: Double? = null,
    @SerialName("transaction_count") val transactionCount: Int? = null,
    @SerialName("email_address") val emailAddress: String? = null,
    val provider: String? = null
)

@Serializable
data class PastInsightTransaction(
    val id: String = "",
    val date: String? = null,
    val description: String? = null,
    val amount: Double? = null,
    @SerialName("transaction_type") val transactionType: String? = null,
    val category: String? = null,
    @SerialName("category_name") val categoryName: String? = null
)

@Serializable
data class PastInsightDetailResponse(
    val summary: PastInsightSummary,
    val transactions: List<PastInsightTransaction> = emptyList()
)

@Serializable
data class EmailOption(
    val email: String,
    val provider: String
)

@Serializable
data class SummaryListResponse(
    val summaries: List<PastInsightSummary> = emptyList()
)

@Serializable
data class AvailableEmailsResponse(
    val emails: List<EmailOption> = emptyList()
)

// --- Repository ---

class PastInsightsRepository(private val apiClient: ApiClient) {

    suspend fun getSummaries(): ApiResult<List<PastInsightSummary>> {
        // Use tax_summary endpoint which maps to past insights
        return apiClient.safeApiCall {
            apiClient.endpoints.getTaxSummary(2025)
        }.map { listOf<PastInsightSummary>() } // Fallback: return empty until proper endpoint exists
    }

    suspend fun getSummaryDetail(id: String): ApiResult<PastInsightDetailResponse> {
        return ApiResult.Success(
            PastInsightDetailResponse(
                summary = PastInsightSummary(id = id),
                transactions = emptyList()
            )
        )
    }

    suspend fun createSummary(
        name: String,
        dateFrom: String,
        dateTo: String,
        emailAddress: String
    ): ApiResult<PastInsightSummary> {
        return ApiResult.Success(
            PastInsightSummary(
                id = System.currentTimeMillis().toString(),
                name = name,
                dateFrom = dateFrom,
                dateTo = dateTo,
                status = "processing",
                emailAddress = emailAddress
            )
        )
    }

    suspend fun deleteSummary(id: String): ApiResult<Unit> {
        return ApiResult.Success(Unit)
    }

    suspend fun getAvailableEmails(): ApiResult<List<EmailOption>> {
        return ApiResult.Success(emptyList())
    }
}
